tpm/tpm.go

```diff
@@ -713,6 +714,106 @@ func OwnerReadSRK(rw io.ReadWriter, ownerAuth digest) ([]byte, error) {
 	return tpmutil.Pack(pk)
 }
 
+// ReadEKCert reads the EKCert from the NVRAM.
+// TCG PC Client Specific Implementation Specification for Conventional BIOS 7.4.4
+func ReadEKCert(rw io.ReadWriter, ownAuth digest) ([]byte, error) {
+	const (
+		certIndex                 = 0x1000f000
+		certTagPCClientStoredCert = 0x1001
+		certTagPCClientFullCert   = 0x1002
+	)
+	offset := uint32(0)
+	data, err := NVReadValue(rw, certIndex, offset, 5, ownAuth)
+	if err != nil {
+		return nil, err
+	}
+	offset = offset + 5
+	buff := bytes.NewReader(data)
+
+	var header struct {
+		Tag      uint16
+		CertType uint8
+		CertSize uint16
+	}
+	var bufSize uint32
+	if err := binary.Read(buff, binary.BigEndian, &header); err != nil {
+		return nil, err
+	}
+
+	if header.Tag != certTagPCClientStoredCert { // TCG_TAG_PCCLIENT_STORED_CERT
+		return nil, fmt.Errorf("invalid certificate")
+	}
+
+	switch header.CertType {
+	case 0: // TCG_FULL_CERT
+		data, err := NVReadValue(rw, certIndex, offset, 2, ownAuth)
+		if err != nil {
+			return nil, err
+		}
+		bufSize = uint32(header.CertSize) + offset
+		offset = offset + 2
+		buff = bytes.NewReader(data)
+
+		var tag uint16
+		if err := binary.Read(buff, binary.BigEndian, &tag); err != nil {
+			return nil, err
+		}
+
+		if tag != certTagPCClientFullCert { // TCG_TAG_PCCLIENT_FULL_CERT
+			return nil, fmt.Errorf("certificate type and tag do not match")
+		}
+	case 1: // TCG_PARTIAL_SMALL_CERT
+		return nil, fmt.Errorf("certType is not TCG_FULL_CERT: currently do not support partial certs")
+	default:
+		return nil, fmt.Errorf("invalid certType: 0x%x", header.CertType)
+	}
+
+	var ekbuf []byte
+	for offset < bufSize {
+		length := bufSize - offset
+		// TPMs can only read so much memory per command so we read in 128byte chunks.
+		// 128 was taken from go-tspi. The actual max read seems to be platform dependent
+		// but cannot be queried on TPM1.2 (and does not seem to appear in any documentation).
+		if length > 128 {
+			length = 128
+		}
+		data, err = NVReadValue(rw, certIndex, offset, length, ownAuth)
+		if err != nil {
+			return nil, err
+		}
+
+		ekbuf = append(ekbuf, data...)
+		offset += length
+	}
+
+	return ekbuf, nil
+}
+
+// NVReadValue consumes an index, offset, and length, and returns the value from that location in NVRAM.
+// See TPM-Main-Part-2-TPM-Structures 19.1.
+func NVReadValue(rw io.ReadWriter, index, offset, len uint32, ownAuth digest) ([]byte, error) {
```

Comment:
Please also add a direct test for NVReadValue. It's useful to be able to quickly distinguish failures of ReadEKCert from failures of NVReadValue.

Reply:
I'm not sure adding a test just for NVReadValue is all that useful without NVDefineSpace and NVWriteValue. There are some predefined NV Indices but the only one that can easily be checked/verified is the EKCert. As such maybe it would be better to add more verbose errors and leave the unit test for when DefineSpace and WriteValue are added?

Reply:
OK, it can wait